File: backend/src/main/java/com/joying/chat/websocket/ChatWebSocketHandler.java
package com.joying.chat.websocket;

import java.security.Principal;
import java.util.Map;
import java.util.concurrent.Executor;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.messaging.handler.annotation.DestinationVariable;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.messaging.simp.SimpMessageHeaderAccessor;
import org.springframework.stereotype.Controller;

import com.joying.chat.broadcast.ChatBroadcaster;
import com.joying.chat.config.KeyOrderedExecutor;
import com.joying.chat.dto.ChatMessageResponse;
import com.joying.chat.dto.SendMessageRequest;
import com.joying.chat.service.ChatPresenceService;
import com.joying.chat.service.ChatService;

/**
 * 웹소켓으로 들어오는 채팅 요청을 받는다.
 *
 * <p>누가 보냈는지는 연결할 때 인증이 남겨 둔 것을 읽는다. 요청마다 다시 확인하지 않는다.
 */
@Controller
public class ChatWebSocketHandler {

	private static final Logger log = LoggerFactory.getLogger(ChatWebSocketHandler.class);

	private final ChatService chatService;
	private final KeyOrderedExecutor messageExecutor;
	private final ChatBroadcaster chatBroadcaster;
	private final ChatPresenceService chatPresenceService;

	public ChatWebSocketHandler(ChatService chatService,
								@Qualifier("chatMessageExecutor") KeyOrderedExecutor messageExecutor,
								ChatBroadcaster chatBroadcaster,
								ChatPresenceService chatPresenceService) {
		this.chatService = chatService;
		this.messageExecutor = messageExecutor;
		this.chatBroadcaster = chatBroadcaster;
		this.chatPresenceService = chatPresenceService;
	}

	/**
	 * 메시지를 보낸다.
	 *
	 * <p>저장과 발행을 요청 스레드에서 하지 않는다. 보낸 사람은 응답을 기다리지 않고,
	 * 결과는 구독으로 돌아온다.
	 *
	 * <p>넓은 조회 풀이 아니라 방 단위로 묶인 실행기에 던진다. 넓은 풀에 던지면 같은
	 * 방의 메시지가 여러 스레드에 흩어져, 번호는 맞는데 받는 쪽에 닿는 순서가 뒤섞인다.
	 * 부하를 넣고 재 보니 200건 중 176번 뒤로 갔다.
	 */
	@MessageMapping("/chat/{chatRoomId}/send")
	public void sendMessage(@DestinationVariable Long chatRoomId,
							@Payload SendMessageRequest request,
							SimpMessageHeaderAccessor headerAccessor) {
		Long memberId = extractMemberId(headerAccessor);

		log.debug("메시지 전송 요청: chatRoomId={}, memberId={}, type={}",
			chatRoomId, memberId, request.getType());

		messageExecutor.execute(chatRoomId, () -> {
			try {
				ChatMessageResponse message = chatService.sendMessage(chatRoomId, memberId, request);
				log.info("메시지 전송 완료: chatRoomId={}, messageId={}, senderId={}",
					chatRoomId, message.getId(), memberId);
			} catch (Exception e) {
				log.error("메시지 전송 실패: chatRoomId={}, memberId={}, error={}",
					chatRoomId, memberId, e.getMessage(), e);
			}
		});
	}

	@MessageMapping("/chat/{chatRoomId}/typing")
	public void sendTypingIndicator(@DestinationVariable Long chatRoomId,
									SimpMessageHeaderAccessor headerAccessor) {
		Long memberId = extractMemberId(headerAccessor);

		chatBroadcaster.toTopic("/topic/chat/" + chatRoomId + "/typing",
			Map.of("chatRoomId", chatRoomId, "memberId", memberId, "isTyping", true));
	}

	@MessageMapping("/chat/{chatRoomId}/read")
	public void markAsRead(@DestinationVariable Long chatRoomId,
						   SimpMessageHeaderAccessor headerAccessor) {
		Long memberId = extractMemberId(headerAccessor);

		chatService.markAsRead(chatRoomId, memberId);

		chatBroadcaster.toTopic("/topic/chat/" + chatRoomId + "/read",
			Map.of("chatRoomId", chatRoomId, "memberId", memberId,
				"readAt", System.currentTimeMillis()));
	}

	/**
	 * 살아 있다는 신호.
	 *
	 * <p>접속 상태는 시간이 지나면 스스로 사라지므로 주기적으로 다시 알려야 한다.
	 * 방 번호를 같이 보내면 그 방을 보고 있다는 것까지 갱신한다.
	 */
	@MessageMapping("/chat/heartbeat")
	public void heartbeat(@Payload(required = false) Map<String, Object> payload,
						  SimpMessageHeaderAccessor headerAccessor) {
		Long memberId = extractMemberId(headerAccessor);

		chatPresenceService.heartbeat(memberId);

		if (payload == null) {
			return;
		}
		Long chatRoomId = toLong(payload.get("chatRoomId"));
		if (chatRoomId != null) {
			chatPresenceService.refreshChatRoomActivity(memberId, chatRoomId);
		}
	}

	private Long toLong(Object value) {
		if (value instanceof Number number) {
			return number.longValue();
		}
		if (value instanceof String text) {
			try {
				return Long.valueOf(text);
			} catch (NumberFormatException e) {
				return null;
			}
		}
		return null;
	}

	/**
	 * 연결할 때 인증이 남겨 둔 사용자 정보를 읽는다.
	 */
	private Long extractMemberId(SimpMessageHeaderAccessor headerAccessor) {
		Principal principal = headerAccessor.getUser();
		if (principal == null) {
			throw new IllegalArgumentException("인증되지 않은 사용자입니다");
		}
		return Long.valueOf(principal.getName());
	}
}

File: backend/src/main/kotlin/com/joying/chat/websocket/ChatWebSocketHandler.kt
package com.joying.chat.websocket

import com.joying.chat.broadcast.ChatBroadcaster

import com.joying.chat.dto.SendMessageRequest
import com.joying.chat.service.ChatPresenceService
import com.joying.chat.service.ChatService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.slf4j.LoggerFactory
import org.springframework.messaging.handler.annotation.DestinationVariable
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.messaging.simp.SimpMessageHeaderAccessor
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Controller

/**
 * WebSocket STOMP 메시지 핸들러
 *
 * 실시간 채팅 메시지 송수신
 */
@Controller
class ChatWebSocketHandler(
    private val chatService: ChatService,
    private val messagingTemplate: SimpMessagingTemplate,
    private val chatBroadcaster: ChatBroadcaster,
    private val chatPresenceService: ChatPresenceService
) {
    private val logger = LoggerFactory.getLogger(ChatWebSocketHandler::class.java)

    /**
     * 메시지 전송
     *
     * SEND /app/chat/{chatRoomId}/send
     * → Redis Pub/Sub를 통해 모든 구독자에게 전달됨
     *
     * @param chatRoomId 채팅방 ID
     * @param request 메시지 내용
     * @param headerAccessor WebSocket 헤더 (인증 정보 추출용)
     */
    @MessageMapping("/chat/{chatRoomId}/send")
    fun sendMessage(
        @DestinationVariable chatRoomId: Long,
        @Payload request: SendMessageRequest,
        headerAccessor: SimpMessageHeaderAccessor
    ) {
        // 인증된 사용자 ID 추출
        val memberId = extractMemberIdFromToken(headerAccessor)

        logger.debug(
            "메시지 전송 요청: chatRoomId={}, memberId={}, type={}",
            chatRoomId,
            memberId,
            request.type
        )

        // 비동기로 메시지 전송 (MongoDB 저장 + Redis Pub/Sub 발행)
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val message = chatService.sendMessage(chatRoomId, memberId, request)

                logger.info(
                    "메시지 전송 완료: chatRoomId={}, messageId={}, senderId={}",
                    chatRoomId,
                    message.id,
                    memberId
                )
            } catch (e: Exception) {
                logger.error(
                    "메시지 전송 실패: chatRoomId={}, memberId={}, error={}",
                    chatRoomId,
                    memberId,
                    e.message,
                    e
                )
            }
        }
    }

    /**
     * 타이핑 중 표시
     *
     * SEND /app/chat/{chatRoomId}/typing
     * → SUBSCRIBE /topic/chat/{chatRoomId}/typing
     *
     * @param chatRoomId 채팅방 ID
     * @param headerAccessor WebSocket 헤더
     */
    @MessageMapping("/chat/{chatRoomId}/typing")
    fun sendTypingIndicator(
        @DestinationVariable chatRoomId: Long,
        headerAccessor: SimpMessageHeaderAccessor
    ) {
        val memberId = extractMemberIdFromToken(headerAccessor)

        logger.debug("타이핑 중: chatRoomId={}, memberId={}", chatRoomId, memberId)

        // 타이핑 상태 브로드캐스트
        val typingEvent = mapOf(
            "chatRoomId" to chatRoomId,
            "memberId" to memberId,
            "isTyping" to true
        )

        chatBroadcaster.toTopic("/topic/chat/$chatRoomId/typing", typingEvent)
    }

    /**
     * 메시지 읽음 처리
     *
     * SEND /app/chat/{chatRoomId}/read
     * → SUBSCRIBE /topic/chat/{chatRoomId}/read
     *
     * @param chatRoomId 채팅방 ID
     * @param headerAccessor WebSocket 헤더
     */
    @MessageMapping("/chat/{chatRoomId}/read")
    fun markAsRead(
        @DestinationVariable chatRoomId: Long,
        headerAccessor: SimpMessageHeaderAccessor
    ) {
        val memberId = extractMemberIdFromToken(headerAccessor)

        logger.debug("읽음 처리: chatRoomId={}, memberId={}", chatRoomId, memberId)

        // 읽음 처리
        chatService.markAsRead(chatRoomId, memberId)

        // 읽음 상태 브로드캐스트
        val readEvent = mapOf(
            "chatRoomId" to chatRoomId,
            "memberId" to memberId,
            "readAt" to System.currentTimeMillis()
        )

        chatBroadcaster.toTopic("/topic/chat/$chatRoomId/read", readEvent)
    }

    /**
     * 온라인 상태 Heartbeat
     *
     * SEND /app/chat/heartbeat
     * 클라이언트가 주기적으로 호출하여 온라인 상태 유지 (30초마다 권장)
     *
     * @param payload chatRoomId를 포함할 수 있는 payload (Optional)
     * @param headerAccessor WebSocket 헤더
     */
    @MessageMapping("/chat/heartbeat")
    fun heartbeat(@Payload payload: Map<String, Any>?, headerAccessor: SimpMessageHeaderAccessor) {
        val memberId = extractMemberIdFromToken(headerAccessor)

        logger.debug("Heartbeat: memberId={}, payload={}", memberId, payload)

        // 온라인 상태 갱신
        chatPresenceService.heartbeat(memberId)

        // chatRoomId가 있으면 채팅방 활성 상태도 갱신
        payload?.get("chatRoomId")?.let { chatRoomId ->
            val chatRoomIdLong = when (chatRoomId) {
                is Number -> chatRoomId.toLong()
                is String -> chatRoomId.toLongOrNull()
                else -> null
            }
            if (chatRoomIdLong != null) {
                logger.debug("Heartbeat: refreshing chat room activity: memberId={}, chatRoomId={}", memberId, chatRoomIdLong)
                chatPresenceService.refreshChatRoomActivity(memberId, chatRoomIdLong)
            }
        }
    }

    /**
     * WebSocket 세션에서 인증된 사용자 ID 반환
     *
     * WebSocketAuthInterceptor가 CONNECT 시점에 쿠키에서 JWT를 추출하여
     * headerAccessor.user에 인증 정보를 저장합니다.
     *
     * @param headerAccessor WebSocket 헤더
     * @return 사용자 ID
     */
    private fun extractMemberIdFromToken(headerAccessor: SimpMessageHeaderAccessor): Long {
        // WebSocketAuthInterceptor에서 설정한 인증 정보 가져오기
        val authentication = headerAccessor.user
            ?: throw IllegalArgumentException("인증되지 않은 사용자입니다")

        // Authentication.name에 memberId가 문자열로 저장되어 있음
        return authentication.name.toLong()
    }
}